packages/std/src/checksum.rs

```diff
@@ -0,0 +1,240 @@
+use core::fmt;
+use core::ops::Deref;
+
+use schemars::JsonSchema;
+use serde::{de, ser, Deserialize, Deserializer, Serialize};
+use sha2::{Digest, Sha256};
+use thiserror::Error;
+
+use crate::{StdError, StdResult};
+
+/// A SHA-256 checksum of a Wasm blob, used to identify a Wasm code.
+/// This must remain stable since this checksum is stored in the blockchain state.
+///
+/// This is often referred to as "code ID" in go-cosmwasm, even if code ID
+/// usually refers to an auto-incrementing number.
+#[derive(JsonSchema, Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Checksum(#[schemars(with = "String")] [u8; 32]);
+
+impl Checksum {
+    pub fn generate(wasm: &[u8]) -> Self {
+        Checksum(Sha256::digest(wasm).into())
+    }
+
+    /// Tries to parse the given hex string into a checksum.
+    /// Errors if the string contains non-hex characters or does not contain 32 bytes.
+    pub fn from_hex(input: &str) -> StdResult<Self> {
+        let mut binary = [0u8; 32];
+        hex::decode_to_slice(input, &mut binary).map_err(StdError::invalid_hex)?;
+
+        Ok(Self(binary))
+    }
+
+    /// Creates a lowercase hex encoded copy of this checksum.
+    ///
+    /// This takes an owned `self` instead of a reference because `Checksum` is cheap to `Copy`.
+    pub fn to_hex(self) -> String {
+        self.to_string()
+    }
+}
+
+impl fmt::Display for Checksum {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for byte in self.0.iter() {
+            write!(f, "{byte:02x}")?;
+        }
+        Ok(())
+    }
+}
+
+impl From<[u8; 32]> for Checksum {
+    fn from(data: [u8; 32]) -> Self {
+        Checksum(data)
+    }
+}
+
+/// Just like Vec<u8>, Checksum is a smart pointer to [u8].
```

Comment:
This is incorrect because Checksum is an array. I'd remove this implementation as the documentation of Deref says this should only be implemented for smart pointers

Reply:
I think our newtypes are a bit of an in-between thing. I think the reason for that rule is so that people don't implement Deref for some struct targeting one of the fields. Imagine having a person that implicitly dereferences to its name. That would be very confusing. But in our case, all the user cares about is the byte array inside the checksum and pretty much the only reason the newtype exists is to provide the correct serialization / deserialization.
That being said, I am also fine with removing it as long as we provide other more explicit ways to get to the array.

Reply:
My concern is not so much the newtype. The newtype in Binary and HexBinary makes sense. But here we don't have a heap allocated vector but an array where the data is stored directly without indirection. So no pointing happening.

Reply:
That is true, but I don't think it matters for the `Deref` trait where the data is stored. It's just a way to tell the compiler how to dereference the type into a different one and that it should do so implicitly. I might be wrong here, but to me the smart pointer requirement is more of a conceptual thing: Is the sole purpose of this type to proxy its inner type without adding invariants about that type on top?
E.g. in the rust book they use `struct MyBox(T)` as an [example to illustrate the Deref trait](https://doc.rust-lang.org/book/ch15-02-deref.html#defining-our-own-smart-pointer).
The [`Cow` type in the rust std lib](https://doc.rust-lang.org/std/borrow/enum.Cow.html) is also an example that, depending on what you put in there, contains data on the stack.

Reply:
I read through the example from the book and I think it's not great. They clearly refer to `Box` but then use an owned `T` in the type for simplicity, which does not make too much sense. What they then try to achieve in the example is dereference of an owned `y: MyBox`, which is going from MyBox to the type it references: `*y`.

In our case an owned `y: Checksum` already provides the full data. I don't see how you would dereference it as this is not a reference.

Now it is right that it could be done. But the Deref docs have a very clear warning which I think applies here as well:

<img width="960" alt="Bildschirmfoto 2023-11-16 um 16 04 03" src="https://github.com/CosmWasm/cosmwasm/assets/2603011/2a7eb120-c668-4ce0-b462-f3d8febb116e">

-----------

I think there are better ways to solve this. E.g. adding `.as_slice` or let instantate2_address accept an `imple AsRef<[u8]>` in the first argument and implement `impl AsRef<[u8]> for Checksum`.